### PR TITLE
Fix randomizer sounds continuing rolling after navigating to different view

### PR DESCRIPTION
**Pull request by ritinae**

### CODE REVIEW

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 1]
```diff
@@ -1,58 +1,111 @@
...
+    setTimeoutHandle(setTimeout(() => rollNext(iteration + 1, restaurants), interpolatedTime))
+  }
+
+  const handleRoll = async (event) => {
     event.preventDefault()
-    setDisableButton(true)
     try {
-      const restaurants = await restaurantService.getAllMatches(filterType, filterCategories, distance)
+      const restaurants = shuffle(await restaurantService.getAllMatches(filterType, filterCategories, distance))
+      //const restaurants = [restaurantsA[0]]
```
Comment (otsha) [inline packages/app/src/components/Randomizer/Randomizer.js]: Remove entirely if not necessary

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 2]
```diff
@@ -1,58 +1,111 @@
...
+    for (let i = array.length - 1; i > 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [array[i], array[j]] = [array[j], array[i]]
+    }
+    return array
   }
 
-  const roll = async (event) => {
+  const easingFunc = (max, min, t) => {
+    //return (1 - t) * min + t * max // Linear interpolation
```
Comment (otsha) [inline packages/app/src/components/Randomizer/Randomizer.js]: Remove entirely if not necessary

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 3]
```diff
@@ -1,58 +1,111 @@
...
+        setRolling(true)
+        setTimeoutHandle(setTimeout(() => rollNext(0, restaurants), maxTimeBetweenRotations))
       } else {
-        setRestaurant({ ...restaurants[0], showMap: true })
+        setRestaurant(restaurants[0])
+        soundService.playFanfare()
       }
-      soundService.playFanfare()
     } catch (errorResponse) {
+      console.log(errorResponse)
```
Comment (otsha) [inline packages/app/src/components/Randomizer/Randomizer.js]: Remove CLOG

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 4]
```diff
@@ -1,58 +1,111 @@
...
+    soundService.playBeep()
+
+    const t = iteration / maxNumberOfRotations
+    let interpolatedTime = easingFunc(maxTimeBetweenRotations, minTimeBetweenRotations, t)
+    if (iterationsRemaining === 1) {
+      interpolatedTime += resultWaitTime
+    }
+
+    setTimeoutHandle(setTimeout(() => rollNext(iteration + 1, restaurants), interpolatedTime))
+  }
```
Comment (otsha) [inline packages/app/src/components/Randomizer/Randomizer.js]: Is there any way to make this shorter / prettier?

**Review by otsha — COMMENTED** [review]
Some comments to be considered.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 5]
```diff
@@ -1,58 +1,111 @@
...
+  const [isRolling, setRolling] = useState(false)
   const [filterType, setFilterType] = useState('some')
   const [filterCategories, setFilterCategories] = useState([])
   const [distance, setDistance] = useState('')
   const [showFilter, setShowFilter] = useState(false)
 
-  const sleep = async (duration) => {
-    return new Promise(r => setTimeout(r, duration))
+  const [timeoutHandle, setTimeoutHandle] = useState()
+  const [nope, setNope] = useState()
```
Comment (skoskipaa) [inline packages/app/src/components/Randomizer/Randomizer.js]: 'Nope' is not very self-explanatory...

Comment (ritinae) [reply packages/app/src/components/Randomizer/Randomizer.js]: `nope` is the hidden *"Easter egg"* feature which re-starts the roll, so reasoning for the name was to make the feature a bit more "hidden" by obfuscating the used state.

For code readability, this is just plain stupid. I'll change these for something more descriptive.

**Review by skoskipaa — COMMENTED** [review]
What Otso said. Otherwise looks good.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 6]
```diff
@@ -1,92 +1,190 @@
...
+    </Button>
+  )
+}
+
+RandomizerButton.propTypes = {
+  onClick: PropTypes.func.isRequired,
+  setError: PropTypes.func.isRequired,
+  disabled: PropTypes.bool,
+  children: PropTypes.node,
+}
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.js]: This sub-component does not provide enough functionality to be considered "required". It is not needed anymore and should be inlined. Leftover/brainfart from some iteration.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 7]
```diff
@@ -1,92 +1,190 @@
...
-        setDistance={setDistance}
-        showFilter={showFilter}
-        setShowFilter={setShowFilter}
+        filterCategories={filter.categories}
+        setFilterCategories={(value) => setFilter({ ...filter, categories: value })}
+        filterType={filter.type}
+        setFilterType={(value) => setFilter({ ...filter, type: value })}
+        distance={filter.distance}
+        setDistance={(value) => setFilter({ ...filter, distance: value })}
+        showFilter={filter.visible}
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.js:148]: I *think* `Filter` should be refactored to just take `filter` and `setFilter` as passing gazillion related props *(and their setters)* as separate properties looks kind of funny.

Prolly out of scope for this PR tho, so did not do it yet.

Comment (ritinae) [reply packages/app/src/components/Randomizer/Randomizer.js:148]: Actually, this thing is even more iffy than what it looks like; `showFilter` is passed in for conditionally rendering an empty fragment in cases where ***this** component does not want to render it*. The filter does not even set the visibility value itself, the condition is purely handled outside it!

I'd propose that if the suggestion about the filter above is implemented, conditional rendering on the `showFilter` prop is moved here from the `Filter`.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 8]
```diff
@@ -1,92 +1,190 @@
...
+import { shuffle } from '../../util/shuffle'
+
 import soundService from '../../services/sound'
 import restaurantService from '../../services/restaurant'
-import Confetti from '../Confetti/Confetti'
 import './Randomizer.css'
 
+const easingFunc = (max, min, t) => {
+  return (max - min) * Math.pow(t, 4) + min
+}
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.js:19]: Couldn't really figure out what to call this thing and where it should reside.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 9]
```diff
@@ -1,92 +1,190 @@
...
+    }
+  }
+
+  const selectRestaurantElement = () => {
+    if (restaurant && restaurant.error) {
+      return <h1 data-testid='randomizer-resultLabel'>{restaurant.error}</h1>
+    }
+
+    return restaurant
+      ? isRolling
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.js]: Should at least have used some clarifying local constants here to give these conditions descriptive names. This whole function likely needs different approach, though.

File: packages/app/src/components/Randomizer/Randomizer.js [hunk 10]
```diff
@@ -1,92 +1,190 @@
...
+
+  const rollsRemaining = maxNumberOfRolls - iteration
+  const isPicky = filter.categories.length > 0
+  const buttonLabel = isRolling
+    ? rollsRemaining > maxNumberOfRolls / 3
+      ? 'Rolling!'
+      : 'Wait for it...'
+    : restaurant
+      ? 'Gimme another one!'
+      : `I'm feeling ${isPicky ? 'lucky' : 'picky'}!`
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.js]: This ~~kind of makes me wish for a nuclear winter.~~ is not the way to handle switching between five different labels. Why isn't **this** in its own component? I think it should be.

File: packages/app/src/components/Randomizer/Randomizer.js (hunk 6 above)
Comment (ritinae) [reply packages/app/src/components/Randomizer/Randomizer.js]: I think the big brain time plan was to move the label handling here to clean up the main component. Should try doing that, probably.

File: packages/app/src/components/Randomizer/Randomizer.test.js [hunk 11]
```diff
@@ -32,59 +34,112 @@ locationService.getLeg.mockResolvedValue({
...
-  const { queryByTestId } = await actRender(<Randomizer />)
-  const buttonElement = queryByTestId('randomizer-randomizeButton')
+test('button for starting the raffle exists', async () => {
+  const { queryByTestId } = await actRender(
+    <Randomizer
+      maxNumberOfRolls={10}
+      maxTimeBetweenRolls={1}
+      minTimeBetweenRolls={0}
+      resultWaitTime={0}
+    />
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/Randomizer.test.js]: Defining these separately for each test is just plain stupid. I should have either defined the values in constants or declared a custom react component `<TestRandomizer />` which just renders the `Randomizer` with these values. Now if any of these props change radically in the future, we need to change every single test in this file.

File: packages/app/src/components/Randomizer/nopeEasterEgg.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+      return true
+    }
+
+    return false
+  }
+
+  return {
+    active: labelVisible,
+    updateAndTryTrigger: update
```
Comment (ritinae) [inline packages/app/src/components/Randomizer/nopeEasterEgg.js:60]: Why are we renaming things? Why can't they be internally called with same names as they are *"exported"*? I need to fix this one, too.

File: packages/app/src/components/Randomizer/Randomizer.js (hunk 4 above)
Comment (ritinae) [reply packages/app/src/components/Randomizer/Randomizer.js]: Yeah, it's ugly. Re-wrote the whole damn thing.

**Review by otsha — APPROVED** [review]
Looks much better now (:

**Review by skoskipaa — APPROVED** [review]
Ok.

**Pull request merged**